addons/udes_product_expiry/models/stock_picking.py

```diff
@@ -0,0 +1,24 @@
+from odoo import api, fields, models,  _
+from odoo.exceptions import ValidationError
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    def action_done(self):
+        """ Override action_done """
+
+        self.validate_expiry_dates()
+        return super(StockPicking, self).action_done()
+
+    def validate_expiry_dates(self):
+        """ Validate expiry dates for all move lines in self based on picking and product config """
+
+        mls = self.mapped('move_line_ids')
+
+        if self.picking_type_id.u_confirm_tracking == 'no':
```

Comment:
Group move lines by picking type and loop over different picking types. It shouldn't happen but makes the code more robust. Also self can contain more than one picking, hence self.picking_typed_id might generate an error.

Reply:
This flag should be u_confirm_expiry_date.